build/Targets/Settings.props

```diff
@@ -140,6 +140,16 @@
     <DeterministicSourceRoot Condition="'$(DeveloperBuild)' != 'true'">/_/</DeterministicSourceRoot>
 
     <EnableSourceLink Condition="'$(DeveloperBuild)' != 'true'">true</EnableSourceLink>
+
+    <!-- It is only intended to automatically run update during dev cycle. However, It will fail the build on CI if the XLF file is not updated.
```

Comment:
❗️ Running a build on a developer machine MUST NOT make changes to files which get committed to source control. Can you confirm that this is the case?

❗️ Adding or modifying a string in the English language resources file SHOULD NOT require a change be made to non-English language resources file(s) during the same commit. Can you confirm that the current approach treats these as two separate actions which are not expected to occur during the same commit?

Reply:
@sharwell Adding or modifying a resource that will be localized _does_ currently cause the related .xlf files (which contain the localized versions) to be updated during build on a **developer** machine. **CI builds** will not modify the files but instead simply fail if the .xlf files are out-of-date.

Adding or modifying an English language resource does not require that the non-English equivalents be provided at the same time; we expect those to be added later by our loc team. The tooling creates placeholders in the .xlf files as needed.

What we need to avoid is a situation where, after merging a change, the .xlf files are completely missing a resource. The workflow and organization of the .xlf files must also meet the needs of our loc team, and to the extent possible be standard across all of our repos. Within those constraints, however, I'm open to suggestions.

Reply:
Different worklfows are possible

# Background
There are a few things that control how/when .xlf files are updated from the source .resx/.vsct/etc. files:

1. UpdateXlfOnBuild: Default false, true means sync the xlf on every build.
2. msbuild /t:UpdateXlf: Update xlf files one time
3. ErrorOnOutOfDateXlf: Default true, which will give an error if there are source changes without corresponding xlf changes

## (1) Workflow of this PR and used in all of our other repos:
* UpdateXlfOnBuild=true on dev builds, false in CI/official
* ErrorOnOutOfDateXlf=true

Dev modifies strings as before, and build updates xlf for them.
Dev submits xlf update with their change. (Dev can't forget this because CI will call them out.)

Repo is ready for loc handoff at any time

## (2) Conservative workflow implied by tooling defaults 
* UpdateXlfOnBuild=false
* ErrorOnOutOfDateXlf=true

Dev gets an error every time they add/remove/modify a string
Dev runs msbuild /t:UpdateXlf to update and eliminate the error.

Build never edits files (not even locally)
Repo is ready for loc handoff at any time


## (3) Workflow that would allow separating Xlf updates from English string changes in different commits

* UpdateXlfOnBuild=false
* ErrorOnOutOfDateXlf=false

Dev works as before and never needs to check in xlf files.

Repo is NOT ready for loc hand-off at any time. 

To get translated, strings must be in the .xlf files as the loc team does not do anything with the source .resx/.vsct files.  So we would need some other automation to update the xlf prior to each hand-off. I have given some thought to this approach and I think it has merit, but we'd have to go build it... If we do, I'm thinking that we would commit the updates to a localization branch and then point the loc team at that branch, which would also eliminate merge conflict pain in the .xlf.

Overall, I think:

(2) is just too disruptive to the normal workflow

(3) would be nicest for devs, but I did not have the bandwidth to go implement it. I'm definitely supportive of it if we have the resources to go do it now.

(1) is a middle ground that has been good enough for the other (admittedly smaller) repos so far. It was something that I could build with a .nupkg alone and no service component. It's also very much like how we've long check in VS-generated Strings.Designer.cs for Strings.resx.

Reply:
I would be more in favor of a new option, call it (4):

* Modify the XLF files in source control to have names of the form **[BaseName].[Language].[SortableDate].xlf**
* During the build, merge all XLF files with the same language, with preference given to the most recent translation unit. This XLF file replaces the ones in source control as the input to the build localization process.
* Apply the update algorithm to the XLF produced in the previous step. Remove all translation units which were not affected from the output process. If any translation units remain, save the output to the intermediate output directory with the form **[BaseName].[Language].*[CurrentSortableDate]*.xlf**.

This approach has many benefits over the ones listed above:

1. The build never errors, because there is no such thing as an "out of date XLF" in source control
1. The process naturally allows for an asynchronous workflow between developers working on the project and localization teams responding to changes
1. Localization teams don't need to deal with strings that are already translated, but can if they want to
1. There are never merge conflicts when updating XLF files in source control, because you never need to modify an existing XLF in source control

Reply:
> The build never errors, because there is no such thing as an "out of date XLF" in source control

Our contract with the localization providers is that we provide .xlf files in the repo with all the strings to be translated and they edit the files with appropriate translations.

Given that, I don't see how splitting into more files prevents the error state where we have strings in the product that are not in any xlf file in the repo. We have to transition out of this state before each handoff. (1) and (2) do that by enforcing the state upon each checkin, (3) does it by sequencing updates with hand-offs.

> The process naturally allows for an asynchronous workflow between developers working on the project and localization teams responding to changes

(3) does this too

> Localization teams don't need to deal with strings that are already translated, but can if they want to
 
This is a non-issue

> There are never merge conflicts when updating XLF 

(3) accomplishes this too

Reply:
> Our contract with the localization providers is that we provide .xlf files in the repo with all the strings to be translated and they edit the files with appropriate translations.

The contract details with the localization providers shouldn't result in workflow decisions that negatively impact daily contributors. Fortunately, since we are using Git, many options are available. If we can't modify the language of the localization contracts, it is easy enough to have the localization providers work from a fork so developers in the main repository are never subjected to the overhead.

> > The process naturally allows for an asynchronous workflow between developers working on the project and localization teams responding to changes
>
> (3) does this too

The process described in (3) results in extraneous commits. I would expect the workflow to involve commits which add/change English language strings in the original resx/etc files, and commits which add/change translated strings in xlf files. Commits which add/change English strings in xlf files are unnecessary overhead which is easily avoided.

> > Localization teams don't need to deal with strings that are already translated, but can if they want to
>
> This is a non-issue

Speaking from experience prior to joining Microsoft this year, this is definitely a benefit for some other companies evaluating localization practices. When these companies look to adopt current practices demonstrated by Microsoft in open source communities, it could be a make-or-break feature of the available tooling.

> > There are never merge conflicts when updating XLF
>
> (3) accomplishes this too

The process described in (3) appears vulnerable to a merge conflict if the following event order occurs:

1. Developer adds a new resource string
2. (Automated) tooling updates the XLF files
3. Translation provider starts the process of localizing resource strings
4. Developer modifies the resource string added in 1
5. (Automated) tooling updates the XLF files
6. Translation provider submits the translated results for 3

Depending on the exact process involved, the merge conflict is likely to happen in step 6 (steps 5 and 6 modified the same file). In the process I described, merge conflicts cannot happen because steps 2 and 5 do not occur, and steps 3/6 always operate on one-off files.

Reply:
> The process described in (3) appears vulnerable to a merge conflict if the following event order occurs:

See this part of (3): "I'm thinking that we would commit the updates to a localization branch and then point the loc team at that branch, which would also eliminate merge conflict pain in the .xlf."

>  I would expect the workflow to involve commits which add/change English language strings in the original resx/etc files, and commits which add/change translated strings in xlf files.

xlf files have both the source and translated text. Localization providers pull the source text from the xlf files and not from the source files. 

This is what a hand-back commit looks like:

https://github.com/dotnet/sdk/commit/f526e32feed46f1c2f190d094ecfc02c37dcf528

It is our responsibility to put up-to-date source text in the .xlf prior to localization hand-off. That's what the updating logic in xliff-tasks does. We can opt to run this less often than on every change, but we cannot hand-off for translation if nothing has updated the .xlf.

> The contract details with the localization providers shouldn't result in workflow decisions that negatively impact daily contributors. Fortunately, since we are using Git, many options are available. [...]

Yes, that's exactly why I said "(3) would be nicest for devs" and why I suggested a branching approach. 

> Speaking from experience prior to joining Microsoft this year, this is definitely a benefit for some other companies evaluating localization practices. 

XLIFF is a standard format that tracks new/modified/translated strings in one file per document per language. There are tools for editing it and they can filter out state==translated for you.